app/services/idempotency/store.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+require 'redis'
+
+module Idempotency
+  class Store
+    # NB the caching vs conflict logic is:
+    # *  re-using the idempotency key with the same request for 10 mins will return the same cached response
+    # *  re-using the idempotency key with the same request after 10 mins to 12 hours will raise a conflict error
+    # *  re-using the idempotency key with a different request will raise a conflict error, for 12 hours
+
+    CACHE_RESPONSE_TTL = 600 # 10 minutes
+    CONFLICT_TTL = 43_200 # 12 hours
+
+    attr_reader :idempotency_key, :conflict_key, :cache_response_key
+
+    def initialize(request)
+      # NB: we need to do a case-insensitive match for IDEMPOTENCY_KEY; rails does not do this automatically if the key contains an underscore
+      @idempotency_key = request.headers.find { |key, _v| key =~ /\AIDEMPOTENCY[\_\-]KEY\Z/i }&.last
+      if @idempotency_key.present?
+        @conflict_key = "conf|#{idempotency_key}"
+        @cache_response_key = "resp|#{idempotency_key}|#{request_hash(request)}"
+      end
+    end
+
+    def get
+      return if @idempotency_key.blank?
```

Comment:
what are the scenarios we might not have an idempotancy key when we get here (as validation errors should stop this at the controller level), I'm wondering if we should raise an error instead of guarding against setting/getting cached responses?

Reply:
maybe we should guard against corrupt/missing requests and responses as well, but again not sure what the scenarios are for those